core/domain/prod_validation_jobs_one_off_test.py

```diff
@@ -16,26 +16,6131 @@
+
+gae_search_services = models.Registry.import_search_services()
+
+USER_EMAIL = 'useremail@example.com'
+USER_NAME = 'username'
+CURRENT_DATETIME = datetime.datetime.utcnow()
+
+(
+    activity_models, audit_models, base_models,
+    collection_models, config_models, email_models,
+    exp_models, feedback_models, file_models,
+    recommendations_models, story_models,
+    user_models,) = (
+        models.Registry.import_models([
+            models.NAMES.activity, models.NAMES.audit, models.NAMES.base_model,
+            models.NAMES.collection, models.NAMES.config, models.NAMES.email,
+            models.NAMES.exploration, models.NAMES.feedback, models.NAMES.file,
+            models.NAMES.recommendations, models.NAMES.story,
+            models.NAMES.user]))
+
+OriginalDatetimeType = datetime.datetime
+
+
+class PatchedDatetimeType(type):
+    """Validates the datetime instances."""
+    def __instancecheck__(cls, other):
+        """Validates whether the given instance is a datatime
+        instance.
+        """
+        return isinstance(other, OriginalDatetimeType)
+
+
+class MockDatetime13Hours(datetime.datetime):
+    __metaclass__ = PatchedDatetimeType
+
+    @classmethod
+    def utcnow(cls):
+        """Returns the current date and time 13 hours behind of UTC."""
+        return CURRENT_DATETIME - datetime.timedelta(hours=13)
+
+
+def run_job_and_check_output(
+        self, expected_output, sort=False, literal_eval=False):
+    """Helper function to run job and compare output."""
+    job_id = self.job_class.create_new()
+    self.assertEqual(
+        self.count_jobs_in_taskqueue(
+            taskqueue_services.QUEUE_NAME_ONE_OFF_JOBS), 0)
+    self.job_class.enqueue(job_id)
+    self.assertEqual(
+        self.count_jobs_in_taskqueue(
+            taskqueue_services.QUEUE_NAME_ONE_OFF_JOBS), 1)
+    self.process_and_flush_pending_tasks()
+    actual_output = self.job_class.get_output(job_id)
+    if literal_eval:
+        actual_output_dict = {}
+        expected_output_dict = {}
+
+        for item in [ast.literal_eval(value) for value in actual_output]:
+            value = item[1]
+            if isinstance(value, list):
+                value = sorted(value)
+            actual_output_dict[item[0]] = value
+
+        for item in [ast.literal_eval(value) for value in expected_output]:
+            value = item[1]
+            if isinstance(value, list):
+                value = sorted(value)
+            expected_output_dict[item[0]] = value
+
+        self.assertEqual(
+            sorted(actual_output_dict.keys()),
+            sorted(expected_output_dict.keys()))
+        for key in actual_output_dict:
+            self.assertEqual(actual_output_dict[key], expected_output_dict[key])
+    elif sort:
+        self.assertEqual(sorted(actual_output), sorted(expected_output))
+    else:
+        self.assertEqual(actual_output, expected_output)
+
+
+def update_datastore_types_for_mock_datetime():
```

Comment:
I am a little surprised by the need to do all of this. Shouldn't you be storing a valid run-of-the-mill datetime which just happens to be a bit older than usual? I can understand mocking some of the datetime functions, but you shouldn't need to monkeypatch App Engine itself, since it should recognize the new datetime a standard Python datetime.

Reply:
Adding a mock datetime results in failure of tests due to missing validator for the mock property ([Ref](https://github.com/arnesson/appengine-python-sdk/blob/master/google_appengine/google/appengine/api/datastore_types.py#L1547)).

I found the fix to monkeypatch AppEngine [here](https://nvbn.github.io/2016/04/14/gae-datastore-freeze-time/). I was not able to find any other way to test invalid datetime fields in the models.

Reply:
Oh! I think maybe you are swapping out the wrong thing. Don't swap out datetime.datetime, swap out the relevant function that's being called in datetime.datetime (e.g. datetime.datetime.utcnow perhaps).

Reply:
Mocking `utcnow` does not work since built-in types in python are immutable. I get the following error:

```
Traceback (most recent call last):
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/prod_validation_jobs_one_off_test.py", line 208, in test_model_with_last_updated_greater_than_current_time
    with self.swap(datetime.datetime, 'utcnow', mock_utcnow):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/ankita/Desktop/opensrc/oppia/core/tests/test_utils.py", line 1647, in swap
    setattr(obj, attr, newvalue)
TypeError: can't set attributes of built-in/extension type 'datetime.datetime'
```